File: cpp/Hash_Table/0454_4Sum_II.cpp
#include <vector>
#include <unordered_map>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
  int fourSumCount(vector<int> &nums1, vector<int> &nums2, vector<int> &nums3, vector<int> &nums4) {
    unordered_map<int, int> map;
    int count = 0;
    for (int num1: nums1) {
      for (int num2: nums2) {
        map[num1 + num2]++;
      }
    }
    for (int num3: nums3) {
      for (int num4: nums4) {
        if (map.find(0 - num3 - num4) != map.end()) count += map[0 - num3 - num4];
      }
    }
    return count;
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0454-4sum-ii.py
class Solution:
    def fourSumCount(self, nums1: [int], nums2: [int], nums3: [int], nums4: [int]) -> int:
        record = dict()
        for i in nums1:
            for j in nums2:
                if i + j in record:
                    record[i + j] += 1
                else:
                    record[i + j] = 1
        count = 0
        for i in nums3:
            for j in nums4:
                key = - i - j
                if key in record:
                    count += record[key]
        return count


nums1 = [1, 2]
nums2 = [-2, -1]
nums3 = [-1, 2]
nums4 = [0, 2]
print(Solution().fourSumCount(nums1, nums2, nums3, nums4))  # 2

nums1 = [0]
nums2 = [0]
nums3 = [0]
nums4 = [0]
print(Solution().fourSumCount(nums1, nums2, nums3, nums4))  # 1
